src/msg/async/AsyncConnection.cc

```diff
@@ -2408,6 +2428,53 @@ void AsyncConnection::handle_ack(uint64_t seq)
   }
 }
 
+void AsyncConnection::DelayedDelivery::do_request(int id)
+{
+  Mutex::Locker l(delay_lock);
+  register_time_events.erase(id);
+  if (delay_queue.empty())
+    return ;
+  utime_t release = delay_queue.front().first;
+  Message *m = delay_queue.front().second;
+  string delay_msg_type = msgr->cct->_conf->ms_inject_delay_msg_type;
+  utime_t now = ceph_clock_now(msgr->cct);
+  if ((release > now &&
+      (delay_msg_type.empty() || m->get_type_name() == delay_msg_type))) {
+    utime_t t = release - now;
+    t.sleep();
+  }
+  delay_queue.pop_front();
+  if (msgr->ms_can_fast_dispatch(m)) {
+    msgr->ms_fast_dispatch(m);
+  } else {
+    msgr->ms_deliver_dispatch(m);
+  }
+}
+
+class C_flush_messages : public EventCallback {
+  std::deque<std::pair<utime_t, Message*> > delay_queue;
+  AsyncMessenger *msgr;
+ public:
+  C_flush_messages(std::deque<std::pair<utime_t, Message*> > q, AsyncMessenger *m): delay_queue(std::move(q)), msgr(m) {}
```

Comment:
would be better if the parameter type of `q` is `std::deque<std::pair<utime_t, Message*> >&&`, otherwise it will be copied when being passed in.